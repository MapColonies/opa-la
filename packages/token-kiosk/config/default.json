{
  "openapiConfig": {
    "filePath": "./openapi3.yaml",
    "basePath": "/docs",
    "rawPath": "/api",
    "uiPath": "/api"
  },
  "telemetry": {
    "tracing": {
      "isEnabled": false
    },
    "shared": {},
    "logger": {
      "level": "info",
      "prettyPrint": false
    }
  },
  "server": {
    "port": 8080,
    "request": {
      "payload": {
        "limit": "1mb"
      }
    },
    "response": {
      "compression": {
        "enabled": true,
        "options": null
      }
    }
  },
  "db": {
    "host": "localhost",
    "port": 5432,
    "username": "postgres",
    "password": "postgres",
    "ssl": {
      "enabled": false,
      "ca": "",
      "key": "",
      "cert": ""
    },
    "database": "token-kiosk"
  },
  "guides": {},
  "qlr": {
    "catalogUrl": "https://catalog.mapcolonies.com",
    "layers": []
  }
}
