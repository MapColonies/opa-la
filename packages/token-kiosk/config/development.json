{
  "auth": {
    "openid": {
      "clientId": "my-local-app",
      "issuerBaseUrl": "http://localhost:8080/realms/my-local-realm",
      "baseUrl": "http://localhost:5173",
      "secret": "sdfsdasdsadsadsadsadsadas",
      "clientSecret": "yfS4DKnZShTLdXUec8Yqi1DAowknKxY3",
      "scopes": "openid profile email",
      "idField": "email",
      "metadataFields": ["email", "email_verified", "name", "given_name", "family_name", "preferred_username"]
    }
  },
  "token": {
    "environment": "prod",
    "expirationDuration": "1w",
    "domains": ["raster"],
    "jwt": {
      "subject": "c2b",
      "issuer": "token-kiosk",
      "algorithm": "RS256"
    },
    "authManager": {
      "baseUrl": "http://localhost:8082",
      "cacheTtlHours": 1
    }
  }
}
