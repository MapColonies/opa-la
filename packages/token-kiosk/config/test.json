{
  "guides": {
    "qgis": "https://example.com/qgis-guide",
    "arcgis": "https://example.com/arcgis-guide"
  },
  "qlr": {
    "catalogUrl": "http://localhost:8085/api/raster/v1",
    "layers": [
      {
        "productType": "Orthophoto",
        "productId": "test_product",
        "displayName": "test product"
      }
    ]
  },
  "db": {
    "host": "postgres",
    "schema": "token_kiosk",
    "username": "postgres",
    "password": "1234",
    "database": "postgres",
    "ssl": {
      "enabled": false
    }
  }
}
