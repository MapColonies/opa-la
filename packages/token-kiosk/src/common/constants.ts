import { readPackageJsonSync } from '@map-colonies/read-pkg';

export const SERVICE_NAME = readPackageJsonSync().name ?? 'unknown_service';
export const DEFAULT_SERVER_PORT = 80;

export const DB_CONNECTION_TIMEOUT = 5000;

export const IGNORED_OUTGOING_TRACE_ROUTES = [/^.*\/v1\/metrics.*$/];
export const IGNORED_INCOMING_TRACE_ROUTES = [/^.*\/docs.*$/];

/* eslint-disable @typescript-eslint/naming-convention */
export const SERVICES = {
  LOGGER: Symbol('Logger'),
  CONFIG: Symbol('Config'),
  TRACER: Symbol('Tracer'),
  METRICS: Symbol('METRICS'),
  AUTH_MANAGER_CLIENT: Symbol('AuthManagerClient'),
  AUTH_MIDDLEWARE: Symbol('AuthMiddleware'),
  PG_POOL: Symbol('PgPool'),
  DRIZZLE: Symbol('Drizzle'),
  HEALTHCHECK: Symbol('HealthCheck'),
} satisfies Record<string, symbol>;
/* eslint-enable @typescript-eslint/naming-convention */
