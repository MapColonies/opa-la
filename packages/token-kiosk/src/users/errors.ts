export class UserIsBannedError extends Error {
  public constructor(userId: string) {
    super(`User with ID ${userId} is banned`);
    this.name = 'UserIsBannedError';
  }
}
