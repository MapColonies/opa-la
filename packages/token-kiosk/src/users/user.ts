import { integer, jsonb, pgSchema, text, timestamp, boolean } from 'drizzle-orm/pg-core';

export const pgDbSchema = pgSchema('token_kiosk');

export const users = pgDbSchema.table('users', {
  id: text('id').primaryKey().notNull(),
  metadata: jsonb('metadata').notNull().default({}),
  createdAt: timestamp('created_at', { withTimezone: true }).notNull().defaultNow(),
  lastRequestedAt: timestamp('last_requested_at', { withTimezone: true }).notNull().defaultNow(),
  token: text('token').notNull(),
  tokenCreationDate: timestamp('token_creation_date', { withTimezone: true }).notNull().defaultNow(),
  tokenExpirationDate: timestamp('token_expiration_date', { withTimezone: true }).notNull(),
  tokenCreationCount: integer('token_creation_count').notNull().default(0),
  isBanned: boolean('is_banned').notNull().default(false),
});

export type User = typeof users.$inferSelect;
export type UserInsert = typeof users.$inferInsert;
