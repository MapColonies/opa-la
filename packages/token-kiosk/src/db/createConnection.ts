import { hostname } from 'node:os';
import { readFileSync, existsSync } from 'node:fs';
import { join } from 'node:path';
import type { commonDbFullV1Type } from '@map-colonies/schemas';
import { drizzle } from 'drizzle-orm/node-postgres';
import { migrate } from 'drizzle-orm/node-postgres/migrator';
import { Pool, type PoolConfig } from 'pg';
import { HealthCheck } from '@godaddy/terminus';
import { promiseTimeout } from '../common/utils';
import { DB_CONNECTION_TIMEOUT } from '../common/constants';
import { users } from '../users/user';

export type DbConfig = PoolConfig & commonDbFullV1Type;

export function createConnectionOptions(dbConfig: DbConfig): PoolConfig {
  const { ssl, ...dataSourceOptions } = dbConfig;
  dataSourceOptions.application_name = `${hostname()}-${process.env.NODE_ENV ?? 'unknown_env'}`;

  const poolConfig: PoolConfig = {
    ...dataSourceOptions,
    user: dbConfig.username,
  };
  if (ssl.enabled) {
    delete poolConfig.password;
    poolConfig.ssl = { key: readFileSync(ssl.key), cert: readFileSync(ssl.cert), ca: readFileSync(ssl.ca) };
  }
  return poolConfig;
}

export async function initConnection(dbConfig: PoolConfig): Promise<Pool> {
  const pool = new Pool(dbConfig);
  await pool.query('SELECT NOW()');
  return pool;
}

export type Drizzle = ReturnType<typeof createDrizzle>;

export function createDrizzle(pool: Pool): ReturnType<typeof drizzle<{ users: typeof users }>> {
  return drizzle(pool, {
    schema: {
      users,
    },
  });
}

export function healthCheck(connection: Pool): HealthCheck {
  return async (): Promise<void> => {
    const check = connection.query('SELECT 1').then(() => {
      return;
    });
    return promiseTimeout<void>(DB_CONNECTION_TIMEOUT, check);
  };
}

// maybe we should test migrations as well. for now, we'll just ignore them
/* istanbul ignore next */
export async function runMigrations(drizzle: Drizzle): Promise<void> {
  const optionalFolders = ['./src/db/migrations', './db/migrations', './migrations'];
  let migrationsFolder: string | null = null;

  for (const folder of optionalFolders) {
    if (existsSync(join(folder, '/meta/_journal.json'))) {
      migrationsFolder = folder;
      break;
    }
  }

  if (migrationsFolder === null) {
    throw new Error('No migrations folder found');
  }

  await migrate(drizzle, { migrationsFolder: migrationsFolder, migrationsSchema: 'token_kiosk' });
}
