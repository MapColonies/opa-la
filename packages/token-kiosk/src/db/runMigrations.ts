import config from 'config';
import { createConnectionOptions, initConnection, createDrizzle, runMigrations } from './createConnection';

(async (): Promise<void> => {
  const pool = await initConnection(createConnectionOptions(config.get('db')));
  await runMigrations(createDrizzle(pool));
  await pool.end();
  console.log('Migrations completed');
})().catch(console.error);
