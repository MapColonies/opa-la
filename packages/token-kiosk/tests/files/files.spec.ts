import { describe, beforeEach, it, expect, beforeAll, afterEach, vi } from 'vitest';
import jsLogger from '@map-colonies/js-logger';
import nock, { abortPendingRequests, cleanAll } from 'nock';
import { trace } from '@opentelemetry/api';
import httpStatusCodes from 'http-status-codes';
import { createRequestSender, RequestSender } from '@map-colonies/openapi-helpers/requestSender';
import type { RequestContext } from 'express-openid-connect';
import type { RequestHandler } from 'express';
import { paths, operations } from '@openapi';
import { getApp } from '@src/app';
import { Drizzle } from '@src/db/createConnection';
import { users } from '@src/users/user';
import { SERVICES } from '@common/constants';
import { initConfig } from '@src/common/config';
import privateKey from '../data/key';
import mockUser from '../data/user';
import { emptyResponse, goodResponse, partialResponse } from './cswResponses';

describe('guides', function () {
  let requestSender: RequestSender<paths, operations>;
  let oidcContext = { user: mockUser } as unknown as RequestContext;
  let drizzle: Drizzle;

  beforeAll(async function () {
    try {
      await initConfig(true);
    } catch (error) {
      console.error('Failed to initialize configuration:', error);
      throw error;
    }
  });

  beforeEach(async function () {
    const middlewareMock: RequestHandler = (req, res, next) => {
      req.oidc = oidcContext as unknown as RequestContext;
      next();
    };

    const [app, container] = await getApp({
      override: [
        { token: SERVICES.LOGGER, provider: { useValue: jsLogger({ enabled: false }) } },
        { token: SERVICES.TRACER, provider: { useValue: trace.getTracer('testTracer') } },
        { token: SERVICES.AUTH_MIDDLEWARE, provider: { useValue: middlewareMock } },
      ],
      useChild: true,
    });

    requestSender = await createRequestSender<paths, operations>('openapi3.yaml', app, { baseUrl: '/api' });
    drizzle = container.resolve<Drizzle>(SERVICES.DRIZZLE);
    nock('http://localhost:8082').get('/key/prod/latest').reply(httpStatusCodes.OK, privateKey);
  });

  afterEach(function () {
    oidcContext = { user: mockUser } as unknown as RequestContext;
    vi.resetAllMocks();
    // Clean up all nock interceptors after each test
    abortPendingRequests();
    cleanAll();
  });

  describe('Happy Path', function () {
    it('should return 200 status code a qlr file', async function () {
      nock('http://localhost:8085').post('/api/raster/v1').reply(httpStatusCodes.OK, goodResponse);

      const res = await requestSender.getFile({ pathParams: { type: 'qlr' } });

      expect(res.status).toBe(httpStatusCodes.OK);
      expect(res).toSatisfyApiSpec();
      expect(res.headers['content-disposition']).toMatch(/attachment; filename="mapcolonies-layers-\d{4}-\d{2}-\d{2}\.qlr"/);
      expect(res.headers['content-type']).toBe('application/xml; charset=utf-8');
      expect(res.text).toMatchSnapshot('qlr-file-content');
    });

    it.skip('should return 200 status code a lyrx file', async function () {
      nock('http://localhost:8085').post('/api/raster/v1').reply(httpStatusCodes.OK, goodResponse);

      const res = await requestSender.getFile({ pathParams: { type: 'lyrx' } });

      expect(res.status).toBe(httpStatusCodes.OK);
      expect(res).toSatisfyApiSpec();
      expect(res.headers['content-disposition']).toMatch(/attachment; filename="mapcolonies-layers-\d{4}-\d{2}-\d{2}\.lyrx"/);
      expect(res.headers['content-type']).toBe('application/json');
      expect(res.body).toMatchSnapshot('lyrx-file-content');
    });
  });

  describe('Bad Path', function () {
    it('should return 401 status code when user is not authenticated', async function () {
      // Mock user as undefined to simulate unauthenticated request
      const unauthenticatedContext = { user: undefined } as unknown as RequestContext;
      oidcContext = unauthenticatedContext;

      const res = await requestSender.getFile({ pathParams: { type: 'qlr' } });

      expect(res).toSatisfyApiSpec();
      expect(res).toHaveProperty('statusCode', httpStatusCodes.UNAUTHORIZED);
    });

    it('should return 403 status code when user is banned', async function () {
      await drizzle.insert(users).values({ id: 'xd@gmail.com', token: '', isBanned: true, tokenExpirationDate: new Date() }).onConflictDoNothing();

      // Mock user as banned by setting up the context to simulate a banned user
      const bannedUserContext = {
        user: {
          ...mockUser,
          email: 'xd@gmail.com',
        },
      } as unknown as RequestContext;

      oidcContext = bannedUserContext;

      const res = await requestSender.getFile({ pathParams: { type: 'qlr' } });

      expect(res).toSatisfyApiSpec();
      expect(res).toHaveProperty('statusCode', httpStatusCodes.FORBIDDEN);
    });
  });

  describe('Sad Path', function () {
    it('should return 500 status code when the catalog service is unavailable', async function () {
      nock('http://localhost:8085').post('/api/raster/v1').reply(httpStatusCodes.SERVICE_UNAVAILABLE);

      const res = await requestSender.getFile({ pathParams: { type: 'qlr' } });

      expect(res).toSatisfyApiSpec();
      expect(res).toHaveProperty('statusCode', httpStatusCodes.INTERNAL_SERVER_ERROR);
    });

    it('should return 500 status code when catalog service returns 404', async function () {
      nock('http://localhost:8085').post('/api/raster/v1').reply(httpStatusCodes.NOT_FOUND);

      const res = await requestSender.getFile({ pathParams: { type: 'qlr' } });

      expect(res).toSatisfyApiSpec();
      expect(res).toHaveProperty('statusCode', httpStatusCodes.INTERNAL_SERVER_ERROR);
    });

    it('should return 500 status code if the catalog service returns 500', async function () {
      nock('http://localhost:8085').post('/api/raster/v1').reply(httpStatusCodes.INTERNAL_SERVER_ERROR);

      const res = await requestSender.getFile({ pathParams: { type: 'qlr' } });

      expect(res).toSatisfyApiSpec();
      expect(res).toHaveProperty('statusCode', httpStatusCodes.INTERNAL_SERVER_ERROR);
    });

    it('should return 500 status code when network error occurs', async function () {
      nock('http://localhost:8085').post('/api/raster/v1').replyWithError('Network error');

      const res = await requestSender.getFile({ pathParams: { type: 'qlr' } });

      expect(res).toSatisfyApiSpec();
      expect(res).toHaveProperty('statusCode', httpStatusCodes.INTERNAL_SERVER_ERROR);
    });

    it('should return 500 if the server returns an empty response', async function () {
      nock('http://localhost:8085').post('/api/raster/v1').reply(httpStatusCodes.OK, '');

      const res = await requestSender.getFile({ pathParams: { type: 'qlr' } });

      expect(res).toSatisfyApiSpec();
      expect(res).toHaveProperty('statusCode', httpStatusCodes.INTERNAL_SERVER_ERROR);
    });

    it('should return 500 status code when the catalog service returns invalid XML', async function () {
      nock('http://localhost:8085').post('/api/raster/v1').reply(httpStatusCodes.OK, '<invalidXml>');

      const res = await requestSender.getFile({ pathParams: { type: 'qlr' } });
      expect(res).toSatisfyApiSpec();
      expect(res).toHaveProperty('statusCode', httpStatusCodes.INTERNAL_SERVER_ERROR);
    });

    it('should return 500 status code when catalog service response is missing some fields', async function () {
      nock('http://localhost:8085').post('/api/raster/v1').reply(httpStatusCodes.OK, partialResponse);

      const res = await requestSender.getFile({ pathParams: { type: 'qlr' } });

      expect(res).toSatisfyApiSpec();
      expect(res).toHaveProperty('statusCode', httpStatusCodes.INTERNAL_SERVER_ERROR);
    });

    it('should return 500 if the catalog service returns a catalog response with no entries', async function () {
      nock('http://localhost:8085').post('/api/raster/v1').reply(httpStatusCodes.OK, emptyResponse);

      const res = await requestSender.getFile({ pathParams: { type: 'qlr' } });
      expect(res).toSatisfyApiSpec();
      expect(res).toHaveProperty('statusCode', httpStatusCodes.INTERNAL_SERVER_ERROR);
    });

    it('should return 500 status code when database connection fails', async function () {
      vi.spyOn(drizzle.query.users, 'findFirst').mockRejectedValue(new Error('Database connection failed'));

      const res = await requestSender.getFile({ pathParams: { type: 'qlr' } });

      expect(res).toSatisfyApiSpec();
      expect(res).toHaveProperty('statusCode', httpStatusCodes.INTERNAL_SERVER_ERROR);
    });
  });
});
