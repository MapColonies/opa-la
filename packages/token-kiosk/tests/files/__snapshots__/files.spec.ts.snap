// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`guides > Happy Path > should return 200 status code a qlr file > qlr-file-content 1`] = `{}`;
